src/core/elf.d

```diff
@@ -0,0 +1,341 @@
+        }
+    }
+
+    dl_phdr_info info;
+
+    void* baseAddress() const
+    {
+        return cast(void*) info.dlpi_addr;
+    }
+
+    const(char)[] name() const
+    {
+        const(char)* cstr = info.dlpi_name;
+
+        // the main executable has an empty name
+        if (cstr[0] == 0)
+            cstr = getprogname();
+
+        import core.stdc.string;
+        return cstr[0 .. strlen(cstr)];
+    }
+
+    /****
+     * Iterates over this object's segments.
+     */
+    int opApply(scope int delegate(ref const Elf_Phdr) @nogc nothrow dg) const
+    {
+        foreach (ref phdr; info.dlpi_phdr[0 .. info.dlpi_phnum])
+        {
+            const r = dg(phdr);
+            if (r != 0)
+                return r;
+        }
+        return 0;
+    }
+
+    bool findSegmentForAddress(in void* address, out const(Elf_Phdr)* result) const
+    {
+        if (address < baseAddress())
+            return false;
+
+        foreach (ref phdr; this)
+        {
+            const begin = baseAddress() + phdr.p_vaddr;
+            if (cast(size_t)(address - begin) < phdr.p_memsz)
+            {
+                result = &phdr;
+                return true;
+            }
+        }
+        return false;
+    }
+}
+
+// -------------------------------
+//  File-based memory-mapped I/O:
+// -------------------------------
+
+struct ElfFile
+{
+@nogc nothrow:
+    static bool open(const(char)* path, out ElfFile file)
+    {
+        file = ElfFile(.open(path, O_RDONLY));
+        return file.isValid();
+    }
+
+    this(int fd)
+    {
+        this.fd = fd;
+        if (fd != -1)
+        {
+            // memory map header
+            this.ehdr = MMapRegion!Elf_Ehdr(fd, 0, Elf_Ehdr.sizeof);
+        }
+    }
+
+    @disable this(this);
+
+    ~this()
+    {
+        if (fd != -1) close(fd);
+    }
+
+    int fd = -1;
+    MMapRegion!Elf_Ehdr ehdr;
+
+    bool isValid() const
+    {
+        return fd != -1 && isValidElfHeader(*ehdr);
+    }
+
+    bool findSectionHeaderByName(const(char)[] sectionName, out ElfSectionHeader header) const
+    {
+        const index = findSectionIndexByName(sectionName);
+        if (index == -1)
+            return false;
+        header = ElfSectionHeader(this, index);
+        return true;
+    }
+
+    size_t findSectionIndexByName(const(char)[] sectionName) const
+    {
+        const stringSectionHeader = ElfSectionHeader(this, ehdr.e_shstrndx);
+        const stringSection = ElfSection(this, stringSectionHeader);
+
+        foreach (i; 0 .. ehdr.e_shnum)
+        {
+            auto sectionHeader = ElfSectionHeader(this, i);
+            auto currentName = getSectionName(stringSection, sectionHeader.sh_name);
+            if (sectionName == currentName)
+                return i;
+        }
+
+        // not found
+        return -1;
+    }
+}
+
+struct ElfSectionHeader
+{
+@nogc nothrow:
+    this(ref const ElfFile file, size_t index)
+    {
+        assert(Elf_Shdr.sizeof == file.ehdr.e_shentsize);
```

Comment:
Tried my upcoming Android commit for #2325 with this pull, but it segfaults here: will look into why. In the meantime, turn this into an `abort` message, just as you did with #2324, so it exits immediately rather than repeatedly trying to throw an exception and then segfaulting. I'm doing the same in my Android pull, replacing asserts with abort messages.

Reply:
This is probably because this module currently only supports *native* ELF files, i.e., corresponding with the host, which is quite a serious limitation.

I don't think aborting here is a good idea; #2324 was different, as the chances for the GC not being initialized yet/already shut down are very high for that code, so throwing in that code is to be avoided.

Reply:
I'm building on Android itself, but it's likely that some of the ELF definitions are different there.

As for aborting here, I will be using this on Android before the GC is initialized, but I didn't check how you're using it.

Reply:
It doesn't matter how I'm using it; if this code is in `core`, the user is supposed to be able to use it too, so I don't want to abort in case the ELF file doesn't match; it might as well just be an attempt of opening the file as e.g. 32-bit ELF file. So I should actually change the assertion to a normal throw.